### PR TITLE
CSI: volume snapshot

### PR DESCRIPTION
**Pull request by tgross**
CSI volume snapshot workflow for #8212. The resize workflows will be implemented in a separate PR.

### CODE REVIEW

File: nomad/csi_endpoint.go [hunk 1]
```diff
@@ -1052,6 +1051,197 @@ func (v *CSIVolume) ListExternal(args *structs.CSIVolumeExternalListRequest, rep
...
+	}
+
+	state, err := v.srv.fsm.State().Snapshot()
+	if err != nil {
+		return err
+	}
+
+	method := "ClientCSI.ControllerCreateSnapshot"
+	var mErr multierror.Error
+	for _, snap := range args.Snapshots {
```
Comment (drewbailey) [inline nomad/csi_endpoint.go:1077]: This is probably a little overly cautious, but if a user sent a malformed http request with a null value in the snapshots array I think we get a nil pointer dereference. Not sure how we usually handle this, may be worth validating in the http layer 
```
{"Snapshots":[null],"Region":"","Namespace":"","SecretID":""}
```

Comment (tgross) [reply nomad/csi_endpoint.go:1077]: Good call. We can't really rely on client code to be correct for sure.

Comment (drewbailey) [reply nomad/csi_endpoint.go:1077]: (same thing for DeleteSnapshot)

File: client/csi_endpoint.go [hunk 2]
```diff
@@ -311,6 +311,135 @@ func (c *CSI) ControllerListVolumes(req *structs.ClientCSIControllerListVolumesR
...
+	// CSI ControllerDeleteSnapshot errors for timeout, codes.Unavailable and
+	// codes.ResourceExhausted are retried; all other errors are fatal.
+	err = plugin.ControllerDeleteSnapshot(ctx, csiReq,
+		grpc_retry.WithPerRetryTimeout(CSIPluginRequestTimeout),
+		grpc_retry.WithMax(3),
+		grpc_retry.WithBackoff(grpc_retry.BackoffExponential(100*time.Millisecond)))
+	if errors.Is(err, nstructs.ErrCSIClientRPCIgnorable) {
+		// if the snapshot was deleted out-of-band, we'll get an error from
+		// the plugin but can safely ignore it
+		c.c.logger.Debug("could not delete snapshot: %v", err)
```
Comment (drewbailey) [inline client/csi_endpoint.go:383]: I'm not sure why `hclogvet` isn't flagging this :( 
```suggestion
		c.c.logger.Debug("could not delete snapshot", "error", err)
```

Comment (tgross) [reply client/csi_endpoint.go:383]: Yeah, that's weird. Looks like there's a couple of these.

File: command/volume_snapshot_delete.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	flags := c.Meta.FlagSet(c.Name(), FlagSetClient)
+	flags.Usage = func() { c.Ui.Output(c.Help()) }
+
+	if err := flags.Parse(args); err != nil {
+		c.Ui.Error(fmt.Sprintf("Error parsing arguments %s", err))
+		return 1
+	}
+	// Check that we get exactly two arguments
+	args = flags.Args()
+	if l := len(args); l != 1 {
```
Comment (drewbailey) [inline command/volume_snapshot_delete.go:69]:
```suggestion
	if l := len(args); l != 2 {
```

File: command/volume_snapshot_delete.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	client, err := c.Meta.Client()
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error initializing client: %s", err))
+		return 1
+	}
+
+	err = client.CSIVolumes().DeleteSnapshot(&api.CSISnapshot{
+		ID:       snapID,
+		PluginID: pluginID,
+		// Secrets: // TODO: how can we safely plumb Secrets through here?
```
Comment (drewbailey) [inline command/volume_snapshot_delete.go:87]: is `Secrets` needed for Delete?

File: nomad/csi_endpoint.go (hunk 1 above)
Comment (tgross) [reply nomad/csi_endpoint.go:1077]: Follow-up, I just double-checked that JSON decoding on this and it's not something we can rely on there unfortunately: https://play.golang.org/p/RYY6w5cQXYC

File: command/volume_snapshot_delete.go (hunk 4 above)
Comment (tgross) [reply command/volume_snapshot_delete.go:87]: Whoops, I never got around to that TODO. Yeah we need to handle that as a command-line flag unfortunately.

File: command/volume_snapshot_delete.go (hunk 3 above)
Comment (tgross) [reply command/volume_snapshot_delete.go:69]: Yikes.

File: nomad/csi_endpoint.go (hunk 1 above)
Comment (drewbailey) [reply nomad/csi_endpoint.go:1077]: 👍 I was able to get a panic by swapping out Snapshots with `[]*api.CSISnapshot{nil}` here https://github.com/hashicorp/nomad/pull/10245/files#diff-ad34cf50edc09fdf1b4573ba6b5ed49a0f9c2810a3ecefccdba6cfe797a5dd78R139

File: command/volume_snapshot_delete.go (hunk 4 above)
Comment (tgross) [reply command/volume_snapshot_delete.go]: Fixed in 5c9bff7

File: command/volume_snapshot_delete.go (hunk 3 above)
Comment (tgross) [reply command/volume_snapshot_delete.go]: Fixed in 5c9bff7

File: nomad/csi_endpoint.go (hunk 1 above)
Comment (tgross) [reply nomad/csi_endpoint.go:1077]: Fixed in 5d7ba9f

File: client/csi_endpoint.go (hunk 2 above)
Comment (tgross) [reply client/csi_endpoint.go]: Fixed in 3a19cd5

**Review by drewbailey — APPROVED** [review]
LGTM!

**Comment by tgross** [comment]
I suspect there'll be a merge conflict once https://github.com/hashicorp/nomad/pull/10283 merges, so I'll wait till that's done and fix it before merging this in.

**Pull request merged**